### ISSUE TITLE
perf(wan): reclaim 81-frame A14B clips on 24 GB — attention, F32 activation scoping, partial block offload

### ISSUE THREAD

**Issue by jamesbrink** [utensils/mold#776]
## Motivation
Split from #769 ("reclaim 81f on 24GB"). A14B's trained clip length is 81 frames; mold ships 53f (fast) / 33f (quality) defaults because the Q5 pair peaks at 23,975 MiB at 53f/832x480 and 81f OOMs on an RTX 4090 (`crates/mold-core/src/manifest.rs:5171-5183`). The ecosystem's standard levers all exist in reach: FlashAttention is already eligible at these call sites, the per-block F32 residual round trips are scopeable, and partial block streaming (Kijai's `blocks_to_swap`, ComfyUI's budgeted lowvram loader) is the community's 81f@720p answer — mold owns the pattern in `flux/offload.rs`, and the Wan DiT's 40 uniform blocks are the friendliest shape for it.

## Evidence — where the VRAM goes at 81f
- **Attention transients.** Wan self/cross attention routes through `crate::attention::attention` (`crates/mold-inference/src/wan/model/transformer.rs:582`); without `flash-attn` the CUDA path chunks queries at 512 (`crates/mold-inference/src/attention.rs:206-218`) but each chunk still materializes `[B*H, 512, N]` score + softmax tensors. At 81f/480p N = 21x30x52 = 32,760 tokens; with 40 heads that is ~1.3 GB per transient at BF16, more on the F32 softmax path. Flash is *already eligible* — the quantized path runs BF16 activations on GPU (`transformer.rs:120-144`) and `flash_is_eligible` accepts head_dim 128 (`attention.rs:284-305`). This is distinct from #736 (shipping flash-attn artifacts) and #734 (dense FLUX/Z-Image call sites): Wan already reaches the dispatcher; this issue is measurement and qualification.
- **Per-block F32 residual round trips.** `WanBlock::forward` casts the full hidden state to F32 for both gated residual adds and back (`transformer.rs:699-722`): several `[1, 32760, 5120]` F32 transients (~671 MB each) per block, x40 blocks.
- **LoRA branch staging** on the fast tier holds ~1.2 GB of F32 adapter weights beside the 10.8 GB expert (`crates/mold-inference/src/wan/lora.rs:325-331`) — BF16 staging returns ~0.6 GB (see the engine-efficiency follow-ups issue).
- **Block streaming is not wired**: `crates/mold-inference/src/batch.rs:422-427` (`block_offload: false`); `flux/offload.rs` is the in-repo precedent. Kijai's wrapper runs 101f 1024x768 on a 4090 with `blocks_to_swap=40` (kijai/ComfyUI-WanVideoWrapper#898); ComfyUI streams a per-budget block remainder with async copy streams (`model_management.py:983-990`). Partial swap of 10-20 of 40 blocks with prefetch is far cheaper than mold's stream-everything FLUX mode (3-5x).

## Proposed work
1. Measure 81f 832x480 q5 with `MOLD_ATTN=flash` on a flash-attn build (RTX 4090): peak VRAM + s/step vs chunked math; sweep `MOLD_ATTN_CHUNK` (512/1024/2048) for non-flash builds.
2. Scope the F32 residual arithmetic in `WanBlock::forward` so at most one full-token F32 transient is alive at a time, verified against the golden-forward fixtures.
3. If arithmetic doesn't close: demand-driven partial offload for the Wan DiT — keep as many blocks resident as the budget allows, stream the remainder with one-block prefetch on a copy stream; compose with the GGUF `QMatMul` source and the LoRA parallel branch; auto-engage under pressure, forceable via `--offload`/`MOLD_OFFLOAD`; flip `block_offload` in the registry. This also unlocks 720p x 81f and the 12-16 GB card tiers.
4. Raise A14B default frames back toward 81 wherever the bespoke admission model (companion P1 issue) proves fit; keep 53/33 as the no-offload fallback. Sync `website/models/wan.md` and the `MOLD_ATTN` reach note in `website/guide/configuration.md:182` (which still says the dispatcher only reaches FLUX/Flux.2/Z-Image — the Wan DiT is a call site now).

## Acceptance criteria
- `wan22-t2v-a14b:q5` renders 81f 832x480 on a 24 GB 4090 under at least one documented configuration, measured peak recorded; 720p 81f completes on 24 GB (any speed) instead of OOMing.
- No behavior change when everything fits; scheduler plan names the offload disposition; measured peaks replace the 53f/33f numbers in manifest comments, CHANGELOG, and wan.md.
- Check the corresponding #769 bullet off when this lands.

**Comment by jamesbrink** [comment]
Baseline measurement on an RTX 4090 (24,564 MiB), from the wan sequence work in #936.

**81 frames no longer fails on a 24 GB card — but not the way this issue proposes.** With auto-chaining wired for wan (#783), `mold run wan22-t2v-a14b:q5 --frames 81 --width 832 --height 480` routes to 2 x 53-frame clips and completes in **367 s**, delivering 106 frames. That is 81+ frames of video reachable on a 4090 today, with no attention, F32-scoping, or offload work.

It is **not** this issue's acceptance criterion and should not be checked off against it. The criterion is one 81-frame denoise inside 24 GB; chaining gives independent clips, and on a T2V checkpoint the seam carries nothing at all (`motion_tail 0`), so the subject can drift across the boundary. A coherent 81-frame clip still needs the work described here.

Two notes for whoever picks this up:

- The `MOLD_ATTN` reach note in `website/guide/configuration.md` is still accurate that the dispatcher covers FLUX/Flux.2/Z-Image, but the Wan DiT is a call site now (`wan/model/transformer.rs`), so item 1's flash measurement is available without new plumbing — it needs a `--features cuda,flash-attn` build.
- Item 2 (scoping the F32 residual round trips in `WanBlock::forward`) is where I would start: the two gated residual adds each hold up to four full-token F32 transients simultaneously (`x32`, `attn32`, the `broadcast_mul` result, and the sum) because Rust keeps every temporary in a chained expression alive to the end of the statement. At 81f/480p that is ~671 MB each. Splitting the statements and dropping eagerly reduces the live set without touching the arithmetic — but it must be verified against the golden-forward fixtures, since the F32 residual is upstream's numerical-stability choice, not an accident.

**Comment by jamesbrink** [comment]
## Measurement first: FlashAttention does *not* close the 81f gap

Built `--features cuda,flash-attn,mp4` and measured on an RTX 4090. Both rows are the **same binary**, only `MOLD_ATTN` varying, so the backend is the only variable.

`wan22-t2v-a14b:q5`, 53f @ 832×480 (the shipped default):

| Backend | Peak VRAM | Wall clock |
| --- | --- | --- |
| `flash` | 21,354 MiB | 75.3 s |
| `math` | 22,250 MiB | 158.4 s |

**Flash buys 896 MiB and 2.1× speed.** The speedup is the headline; the memory saving is much smaller than this issue assumed ("~1.3 GB per transient", attention as the dominant VRAM lever). It is not.

### Why that matters for the 81f target

81 frames is currently refused at **~27.7 GB against ~24.8 GB usable** — a 2.9 GB gap, not the wholesale OOM described here. (The per-token timestep compaction from #802 item 1 already moved this.)

The admission model is faithful at these shapes. Its calibration points give a slope of 0.4854 MiB/token (17f = 16,214 MiB and 53f = 23,030 MiB, i.e. 7,800 and 21,840 tokens), and extrapolating to 81f's 32,760 tokens gives 28,331 MiB — which matches the 27.7 GB the estimator actually reports.

Flash's measured saving is 896 MiB over 21,840 tokens = 0.041 MiB/token. **Extrapolated** to 81f that is ~1,343 MiB, landing 81f at roughly 27.0 GB. Still above the 24.8 GB ceiling.

So: item 1 is done and the answer is that **it is a speed lever, not the memory lever**. Reaching 81f needs item 3 (demand-driven partial block offload). Item 2 (F32 residual scoping) is bounded too — the gated residual add is computed in FP32 to match upstream's `with amp.autocast(dtype=torch.float32)` around `x = x + y * e[2]`, so it cannot be removed without deviating from the reference; the most it can give is dropping one of the simultaneously-live F32 temporaries.

### Consequence for the estimator

Do **not** zero the score term for flash builds. The 2.26 measured slope is a fit over the whole derived tensor sum, so the score term is not an independent physical quantity — removing it would cut ~4 GB from the estimate where reality frees 0.9 GB, and under-charging is the direction that OOMs mid-denoise. The current behaviour (documented as over-estimating on flash builds, in the safe direction) should stay until someone recalibrates the whole model against flash measurements.

Also worth noting: `flash-attn` ships in no release artifact, so any 81f-on-flash configuration would be source-build-only.

**Comment by jamesbrink** [comment]
## Item 2 landed and measured; item 3 is blocked on a type-erasure problem worth naming

### Item 2 — gated residual F32 scoping: done, 832 MiB

`x + y * gate` was one expression, so four full-clip F32 temporaries were alive at once
(671 MiB each at 81f/832x480, two sites x 40 blocks). The add is elementwise, so it now
runs in 4096-token slices — exact, pinned bit-for-bit against the old whole-tensor form at
token counts either side of the slice width and with both the scalar and TI2V per-token
gate shapes.

**RTX 4090, `wan22-t2v-a14b:q5`, 53f/832x480, sampling `nvidia-smi`: 22,218 -> 21,386 MiB.**

### The arithmetic does not close 81f, and here is the number

Two measured points with the new code (CFG off, 832x480):

| Frames | Tokens | Peak |
| -----: | -----: | ---: |
| 17 | 7,800 | 16,074 MiB |
| 53 | 21,840 | 21,354 MiB |

0.376 MiB/token over a 13,141 MiB intercept, so **81f extrapolates to ~25,461 MiB against a
24,564 MiB card — still ~900 MiB over.** Item 3 is genuinely required, as this issue
anticipated.

### Re-fitting the estimate exposed a latent OOM

Worth flagging on its own. Re-fitting the slope for the new formula (2.26 -> 2.14, derived
per-token 225,280 -> 184,320 B) made admission *accept* 81f. The cause was not the new
number: `WAN_FIXED_WORKSPACE_BYTES` was 0 with the whole fixed cost folded into the slope,
and a difference-based fit cancels the intercept, so applying it per token re-charges that
intercept once for every token. The old too-steep slope covered the gap; the correct slope
did not. Fixed by charging the fitted 2,301 MiB intercept — the model now predicts 21,354
MiB where 21,322 was measured, and 81f is refused again for the right reason.

`WAN_CFG_RESIDENT_BYTES` also fell 945 -> 256 MiB: the uncond pass used to leave full-clip
F32 transients alive and now cannot. Re-measured, A14B's CFG delta is **-32 MiB** (noise);
only the 1.3B still shows one (+256 MiB). All four calibration points re-measured.

### Item 3 — the blocker is `WanLinear`, not the offload logic

`flux/offload.rs` is the cited precedent and it works because FLUX's blocks are concrete
typed structs — `DoubleBlock`/`SingleBlock` with a hand-written `to_device` per weight,
~1,300 lines. Wan cannot reuse that shape as-is:

```rust
pub(crate) type WanLinear = Arc<dyn Module + Send + Sync>;
```

Every projection in a Wan block is type-erased behind `dyn Module`. There is no
`to_device`, and no way to write one generically over an opaque trait object. So item 3
needs one of:

1. **De-erase `WanLinear` into a concrete enum** with a `to_device` arm per variant —
   BF16, GGUF `QMatMul`, FP8 (`wan/model/fp8.rs`), and the LoRA parallel branch. This is
   the right answer and composes with the GGUF source and LoRA branch the issue requires,
   but it touches every construction site in the family.
2. **Reload each swapped block from the source file** — much smaller, but it puts a cold
   disk read back in the step loop, which is exactly what #802 item 3's prefetch was
   added to remove.

I stopped rather than half-landing this: a partial block offload that mishandles the
quantized path or the LoRA branch is a correctness hazard in the memory system, not just a
perf miss.

### A cheaper route to the acceptance criterion, if wanted

The criterion is 81f "under at least one documented configuration". FlashAttention measures
896 MiB at 53f (#776 item 1), which scales to ~1,344 MiB at 81f — enough to put the real
peak near 24,117 MiB, inside the card. What blocks it is admission, not the hardware: the
estimator has no flash-aware slope. Note the fix is **not** zeroing the score term — under
the corrected model that cuts 5.7 GB where reality frees 1.3 GB, because the measured
multiplier applies to that term too. It needs its own two-point fit on a `flash-attn`
build, same protocol as above. That is a contained measurement cycle rather than the
architectural change item 3 needs.

**Comment by jamesbrink** [comment]
## Item 3 design: the type-erasure blocker is real but de-erasure is the wrong fix

Followed up on the `WanLinear = Arc<dyn Module + Send + Sync>` blocker I posted above. The
unknown that actually gates the design was whether a **quantized** block can change device
losslessly — if it cannot, GGUF offload is off the table regardless of how the types are
shaped. It can:

- `QTensor::data() -> Cow<[u8]>` hands back the raw quantized bytes
  (`candle-core/src/quantized/mod.rs:803`).
- `ggml_file::qtensor_from_ggml(dtype, raw_data, dims, device)` is public and rebuilds a
  `QTensor` on **any** device from those bytes (`ggml_file.rs:138`).

So GPU->CPU->GPU is a byte round trip, not a dequantize/requantize. `quantize_onto` is the
trap to avoid here — it takes an F32 source and would re-quantize, which is lossy.

That makes two implementations viable, and **neither requires de-erasing `WanLinear`**:

**Option A — byte-move.** A `WanModule` trait (supertrait of `Module`, so every existing
call site is untouched) with `to_device` + `weight_bytes`, implemented per concrete linear:
`candle_nn::Linear`, the GGUF `QMatMul` one, the fp8 one (`wan/model/fp8.rs`), plus
delegating impls for `CastBoundary` and the LoRA parallel branch. Keeps a host-RAM copy of
the parked blocks — fastest swap, but ~10.8 GB of host RAM for a full A14B expert.

**Option B — rebuild from the var-builder.** Keep the `WanWeights` alive and rebuild a
block on the target device with the existing, already-tested `WanBlock::load` path
(`VarBuilder::set_device` exists, `var_builder.rs:245`). No new per-type code at all and no
host-RAM copy, because the GGUF/safetensors source is mmap'd — a swap is a page-cache read
plus an upload. The cost is that it is a file read in the step loop, which is exactly what
#802 item 3's partner-expert prefetch was added to hide; the same prefetch pattern applies
directly.

Option B is far less code and reuses a construction path that is already correct for all
three weight sources and the LoRA branch — which is the part most likely to be got subtly
wrong in Option A. Option A is worth it only if measurement shows the page-cache read
dominates, and that is measurable before committing to it.

Either way the remaining work is the driver, not the plumbing: keep as many blocks resident
as the budget allows, stream the remainder with one-block prefetch, auto-engage under
pressure, honour `--offload`/`MOLD_OFFLOAD`, flip `block_offload` in
`crates/mold-inference/src/batch.rs`, and name the disposition in the scheduler plan.

Sizing, from the calibration re-fit in the comment above: an 81-frame A14B q5 render
extrapolates to ~25,461 MiB against a 24,564 MiB card, and one of the 40 blocks is ~270 MiB
at Q5. **Parking ~8 blocks covers the ~900 MiB shortfall with margin**, which is a long way
short of the stream-everything FLUX mode and should cost far less than its 3-5x.

Not started in this session — flagging the settled design rather than half-landing a
memory-system change.

**Comment by jamesbrink** [comment]
Closing as completed — PR #995 explicitly closed the remaining items (3, 4, and both acceptance criteria):
- Attention measured (flash: 896 MiB / 2.1x on 4090, not the memory lever); #940 corrected the `MOLD_ATTN` reach note
- F32 activation scoping: #949 (4096-token chunks in the gated residual, −832 MiB, admission re-fitted)
- Partial block offload: #954 + #995 (`wan/block_offload.rs`, `MOLD_WAN_OFFLOAD_BLOCKS`, plan disposition `wan_block_offload`)
- Defaults raised on measurement: Q5/Q4 → 81f, Q8 → 73f, fp8 → 45f (`manifest.rs`); 720p ceilings documented in `website/models/wan.md`
Unmeasured leftover (not in acceptance): the `MOLD_ATTN_CHUNK` 512/1024/2048 sweep for non-flash builds.
